### ISSUE TITLE
Broken Code in check_dns Version 2.3.alpha while working with multiple ip addresses 

### ISSUE THREAD

**Issue by tuxracer1337** [monitoring-plugins/monitoring-plugins#1648]
Hello guys,

we tried the 2.3.alpha Version of check_dns and now not all of our checks with check_dns will work.

# This host has Version 2.2

```
host1:~> /usr/lib/nagios/plugins/check_dns -H foo.bar.net -a 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44 -t 10
DNS OK: 0.009 seconds response time. foo.bar.net returns 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44|time=0.009254s;;;0.000000
```

# This host has Version 2.3.alpha

```
host2:~> /usr/lib/nagios/plugins/check_dns -H foo.bar.net -a 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44 -t 10
DNS CRITICAL - expected '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
```

For us it looks like the issue come with one of the following commits from @DerDakon with changes around multiple ip addresses. But we're not sure. 

https://github.com/monitoring-plugins/monitoring-plugins/commit/a03068743f3694cbdbe84bb6e802a41f270cc105#diff-dc3d9c09b035e2ce86ff1fd16737a3bc79683e28f5c985767176320783ac2871
https://github.com/monitoring-plugins/monitoring-plugins/commit/fd9a7d2e00b2acf66ec9aa48d7f51daa817f3457#diff-dc3d9c09b035e2ce86ff1fd16737a3bc79683e28f5c985767176320783ac2871

For us it's important to fix this befor the alpha version goes to stable branch.

If you require more information, let me know what you're missing.

cheers

Andreas

**Comment by DerDakon** [comment]
That looks indeed suspicious. Could you please try passing only  one address in -a, and check for each address individually if you can find anything that doesn't work? Can you check if adding '-L' with the full address list makes a difference? I guess you want this anyway, i.e. it requires an exact match of all addresses.

**Comment by tuxracer1337** [comment]
HI @DerDakon,

# Test with -L as additional parameter
```
host2:~ # /usr/lib/nagios/plugins/check_dns -H foo.bar.net -a 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44 -t 10 -L
DNS CRITICAL - expected '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
```
Same error as without -L

# Test with a single ip for parameter -a
```
host2:~ # for i in 10.10.94.92 10.10.94.93 10.10.39.33 10.10.39.34 10.10.39.43 10.10.39.44; do /usr/lib/nagios/plugins/check_dns -H foo.bar.net -a $i -t 10; done
DNS OK: 0.009 seconds response time. foo.bar.net returns 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44|time=0.008649s;;;0.000000
DNS OK: 0.007 seconds response time. foo.bar.net returns 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44|time=0.007076s;;;0.000000
DNS OK: 0.007 seconds response time. foo.bar.net returns 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44|time=0.007020s;;;0.000000
DNS OK: 0.007 seconds response time. foo.bar.net returns 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44|time=0.007239s;;;0.000000
DNS OK: 0.006 seconds response time. foo.bar.net returns 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44|time=0.006396s;;;0.000000
DNS OK: 0.006 seconds response time. foo.bar.net returns 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44|time=0.006293s;;;0.000000
```
That works, but this is not the way we want to use.

# Test with a single ip for parameter -a and additional -L parameter
```
host2:~ # for i in 10.10.94.92 10.10.94.93 10.10.39.33 10.10.39.34 10.10.39.43 10.10.39.44; do /usr/lib/nagios/plugins/check_dns -H foo.bar.net -a $i -t 10 -L; done
DNS CRITICAL - expected '10.10.94.92' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
DNS CRITICAL - expected '10.10.94.93' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
DNS CRITICAL - expected '10.10.39.33' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
DNS CRITICAL - expected '10.10.39.34' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
DNS CRITICAL - expected '10.10.39.43' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
DNS CRITICAL - expected '10.10.39.44' but got '10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44'
```
The error here is similar to that when executing with a list.

I hope this will help you to find a solution

**Comment by DerDakon** [comment]
Could you please run:

```shell
strace -f -s 256 -o bug1648.log /usr/lib/nagios/plugins/check_dns -H foo.bar.net -a 10.10.94.92,10.10.94.93,10.10.39.33,10.10.39.34,10.10.39.43,10.10.39.44 -t 10
```

And then just send me the output? Direct email is fine.

**Comment by tuxracer1337** [comment]
@DerDakon 
I was sending an email with the requested strace output to eike@sf-mail.de

**Comment by DerDakon** [comment]
Ok, the error output is a bit misleading. The expected part is your string, and the thing it got is the concat of all the addresses it found. The point is: it expects one address, as you passed "-a" only once. Pass every address as a single "-a" argument, then it will work.

**Comment by tuxracer1337** [comment]
@DerDakon 

> pass every address as a single "-a" argument, then it will work.

This is not a solution for me. I not only have one check in an environment that would be affected, I have some environments with many such checks, so the effort is not insignificant.
Especially when I'm using a branch that doesn't include the changes you made, everything seems to work for me.

**Comment by DerDakon** [comment]
This has only ever worked by accident. My use case is: I have the normal host address and a custom variable holding the IPv6 address per host. Generating the proper '-a' argument was impossible as it needed to be sorted alphabetically to get the right argument for check_dns. This is the reason I changed this.

We now can add additional code to split one "-a" at commas into multple arguments, or you just do

```shell
sed '/check_dns/s/,/ -a /g' -i /etc/nagios/services/dns.cfg
```

Or something similar. I'm pretty sure that the changed config would also work with the old check_dns. I could provide the split code as an additional patch, but I still think you gain more flexibility and correctness if you just fix your config files.

**Comment by tuxracer1337** [comment]
Let me explain why it is more as a sed.

Some of our customer is using a CMDB for managing icinga checks in combination with the icinga director.
These tools would be to rewriten too,  so they can no longer provide a list.

And sorry, but "This has only ever worked by accident." isn't the truth. As we can see in the help page below, it works as designed befor you made any changes.

```
~ # /usr/lib/nagios/plugins/check_dns -h
check_dns v2.2 (monitoring-plugins 2.2)
Copyright (c) 1999 Ethan Galstad <nagios@nagios.org>
Copyright (c) 2000-2008 Monitoring Plugins Development Team
        <devel@monitoring-plugins.org>
 
This plugin uses the nslookup program to obtain the IP address for the given host/domain query.
An optional DNS server to use may be specified.
If no DNS server is specified, the default server(s) specified in /etc/resolv.conf will be used.
 
 
Usage:
check_dns -H host [-s server] [-a expected-address] [-A] [-t timeout] [-w warn] [-c crit]
 
Options:
-h, --help
    Print detailed help screen
-V, --version
    Print version information
--extra-opts=[section][@file]
    Read options from an ini file. See
    https://www.monitoring-plugins.org/doc/extra-opts.html
    for usage and examples.
-H, --hostname=HOST
    The name or address you want to query
-s, --server=HOST
    Optional DNS server you want to use for the lookup

>>> -a, --expected-address=IP-ADDRESS|HOST
>>>     Optional IP-ADDRESS you expect the DNS server to return. HOST must end with
>>>     a dot (.). This option can be repeated multiple times (Returns OK if any
>>>     value match). If multiple addresses are returned at once, you have to match
>>>     the whole string of addresses separated with commas (sorted alphabetically).**

-A, --expect-authority
    Optionally expect the DNS server to be authoritative for the lookup
-w, --warning=seconds
    Return warning if elapsed time exceeds value. Default off
-c, --critical=seconds
    Return critical if elapsed time exceeds value. Default off
-t, --timeout=INTEGER
    Seconds before connection times out (default: 10)
 
Send email to help@monitoring-plugins.org if you have questions regarding
use of this software. To submit patches or suggest improvements, send email
to devel@monitoring-plugins.org
```

At next code block you can see your recommended usage of multiple "-a" parameters doesn't work in the version 2.2:

```
~ # /usr/lib/nagios/plugins/check_dns –H foo.bar.net -a 10.10.94.92 -a 10.10.94.93 -a 10.11.39.33 -a 10.11.39.34 -a 10.11.39.43 -a 10.11.39.44
DNS CRITICAL - expected '10.10.94.92; 10.10.94.93; 10.11.39.33; 10.11.39.34; 10.11.39.43; 10.11.39.44' but got '10.10.94.92,10.10.94.93,10.11.39.33,10.11.39.34,10.11.39.43,10.11.39.44'
```

Due to your changes, the CHECK is no longer downward compatible. Therefore, I would like to ask you to remove your changes from the code and then to enter them so that it is also downward compatible.

**Comment by DerDakon** [comment]
You could provide some automated tests for this ;)

**Comment by sni** [comment]
thanks for the investigation, should be fixed with #1649
